be/src/storage/lake/lake_persistent_index.cpp

```diff
@@ -248,23 +248,49 @@ Status LakePersistentIndex::try_replace(size_t n, const Slice* keys, const Index
     return Status::OK();
 }
 
-std::unique_ptr<sstable::Iterator> LakePersistentIndex::prepare_merging_iterator() {
+Status LakePersistentIndex::prepare_merging_iterator(
+        const TabletMetadata& metadata, TxnLogPB* txn_log,
+        std::vector<std::shared_ptr<PersistentIndexSstable>>* merge_sstables,
+        std::unique_ptr<sstable::Iterator>* merge_iter_ptr) {
     sstable::ReadOptions read_options;
     std::vector<sstable::Iterator*> iters;
+    DeferOp free_iters([&] {
+        if ((*merge_iter_ptr) == nullptr) {
+            // free iters if build merge iter fail
+            for (sstable::Iterator* each : iters) {
+                delete each;
+            }
+        }
+    });
+
     auto max_compaction_versions = config::lake_pk_index_sst_max_compaction_versions;
     iters.reserve(max_compaction_versions);
-    for (const auto& sstable : _sstables) {
-        sstable::Iterator* iter = sstable->new_iterator(read_options);
+    std::stringstream ss_debug;
+    for (const auto& sstable_pb : metadata.sstable_meta().sstables()) {
+        // build sstable from meta, instead of reuse `_sstables`, to keep it thread safe
+        ASSIGN_OR_RETURN(auto rf,
+                         fs::new_random_access_file(_tablet_mgr->sst_location(_tablet_id, sstable_pb.filename())));
+        auto* block_cache = _tablet_mgr->update_mgr()->block_cache();
+        if (block_cache == nullptr) {
+            return Status::InternalError("Block cache is null.");
+        }
+        auto merge_sstable = std::make_shared<PersistentIndexSstable>();
+        RETURN_IF_ERROR(merge_sstable->init(std::move(rf), sstable_pb, block_cache->cache()));
```

Comment:
Why `sstables` are needed to be cloned? Is it because `SeekableInputStream` is not thread-safe?

Reply:
No, we need to make sure iterator release before sstable.